Plugins: Refactor kvstore usage in signing keys and angular patterns

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Refactoring work that introduces package `cachekvstore`.

It's a wrapper around a kvstore that also allows to set/get last updated time for the whole store, which can be used to cache values and check if they are stale.

This puts the shared code between the angular patterns store and signing keys rotation store in one place.

(in both cases, we have to periodically fetch values from GCOM, store them in the database and re-fetch them once they "expire")

**Why do we need this feature?**

Refactoring, cleanup.

**Who is this feature for?**

Plugins platform squad.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

I would recommend checking the new `cachekvstore` package first, which has the interfaces and implementations.

Then you can check out usage in `angularpatternsstore` and `keystore`.

Also, I wrote docstrings for everything and a lot of tests, so don't get scared by the number of lines changed :)

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

xnyo committed:
Initial refactoring work for plugins kvstore

xnyo committed:
Replace implementations for keystore and angularstore

xnyo committed:
Cleanup

xnyo committed:
add interface check

xnyo committed:
lint

xnyo committed:
fix storeKeyGetter not being called in namespacedstore set

xnyo committed:
Fix tests

xnyo committed:
Comments

xnyo committed:
Add tests

xnyo committed:
Fix invalid cap in ListKeys when store is empty

xnyo committed:
Update docstrings

xnyo committed:
Add setLastUpdatedOnDelete

xnyo committed:
Renamed DefaultStoreKeyGetterFunc, add TestDefaultStoreKeyGetter

xnyo committed:
Sort imports

xnyo committed:
Merge branch 'main' into giuseppe/kvstore-refactoring

marefr commented on pkg/services/pluginsintegration/angularpatternsstore/store.go:
```diff
@@ -24,52 +18,16 @@ const (
-		return time.Time{}, nil
+		SingleKeyStore: cachekvstore.NewSingleKeyNamespacedStore(
+			kv, kvNamespace, keyPatterns,
+			cachekvstore.WithLastUpdatedKey(keyLastUpdated),
```
Nit. This is the same as `cachekvstore.DefaultLastUpdatedKey`. Maybe `cachekvstore` should only support `"last_updated"` to simplify?

marefr reviewed (commented):
Looks good in general. Great work. It might feel a bit complicated/adding lots of abstractions though compared to earlier where we had two rather simple/straight forward implementations. I'm all in for code-reuse, but maybe not until it really hurts to maintain these implementations? I'll bring it up with the rest of the team for discussion next week and we can decide how to move forward

andresmgot commented on pkg/services/pluginsintegration/cachekvstore/namespacedstore.go:
```diff
@@ -0,0 +1,166 @@
+// It uses the Marshal method to marshal the value before storing it.
+// This means that the value to store can implement the Marshaler interface to control how it is stored.
+func (s *NamespacedStore) Set(ctx context.Context, key string, value any) error {
+	valueToStore, err := Marshal(value)
```
I think that you can omit all the `Marshaler` implementation if you just do the `Marshal` function here. You don't need to expose either:

```go
func marshal(value any) (string, error)
```

I don't think the customization for this Marshal method is needed.

andresmgot commented on pkg/services/pluginsintegration/angularpatternsstore/store.go:
```diff
@@ -1,18 +1,12 @@
-	Get(ctx context.Context) (string, bool, error)
-	Set(ctx context.Context, patterns any) error
-	GetLastUpdated(ctx context.Context) (time.Time, error)
+	cachekvstore.SingleKeyStore
```
can you make this a NamespacedStore? So you don't need two different implementations for the same store.

andresmgot reviewed (commented):
Thanks @xnyo ! I agree with @marefr, you are adding a lot of flexibility here but that has the price that it's more complex. I think we can highly simplify this removing this customization logic. We should have a single interface exposed in the cachekvstore and both the angular detector and the key retriever should use it the same way (no matter if there are internal changes of functionality, as far those changes are not breaking anything).

marefr commented:
I also checked if there's any potential other usage outside of plugins that could make use of the abstractions added here. The one I had in mind was usage stats, but seems a bit different since it writes two keys to the kv store where's one is last_sent and the other some identifier, but they're not written at the same time. So probably not a good fit. 

https://github.com/grafana/grafana/blob/050b9d9a856ba79ef2c896ec8e644debe1990160/pkg/infra/usagestats/service/service.go#L63
https://github.com/grafana/grafana/blob/050b9d9a856ba79ef2c896ec8e644debe1990160/pkg/infra/usagestats/service/usage_stats.go#L155

xnyo committed:
PR review: removed last_updated key

xnyo committed:
PR review: Removed setLastUpdatedOnDelete

xnyo committed:
Re-added relevant tests

xnyo committed:
PR review: Removed SingleKeyStore

xnyo committed:
PR review: Removed custom marshaling support

xnyo committed:
Renamed marshaler.go to marshal.go

xnyo commented:
Thanks for the feedback! :smile: 
I have simplified things as requested:
- Removed the ability to change the last updated key (changed to const, `last_updated`)
- Removed the ability to enable/disable updating last updated time on delete (changed to be always `true`)
- Removed custom marshaler support and `JSONMarshaler`, now it's JSON by default, or it can be set by implementing `fmt.Stringer`
- Removed `SingleKeyStore` and implemented it as a `NamespacedStore` instead for angular patterns, so we do not require two implementations

marefr commented on pkg/services/pluginsintegration/cachekvstore/cachekvstore.go:
Are these interfaces used now/still needed? Seems like `*cachekvstore.NamespacedStore` are used by the consumers. If not, would remove them.

andresmgot replied:
+1 consumers should just use `plugins.KeyStore`. This package should just contain the implementation.

xnyo replied:
Yes you're right, they aren't required anymore. I have removed them

marefr commented on pkg/services/pluginsintegration/cachekvstore/marshal.go:
Nit. Perhaps put this in the namepacedstore.go

marefr reviewed (approved):
LGTM with a couple nits/comments

andresmgot commented on pkg/services/pluginsintegration/angularpatternsstore/store.go:
```diff
@@ -2,74 +2,42 @@ package angularpatternsstore
-	}
-	return t, nil
+// Set stores the given angular patterns in the underlying cachekvstore.
+// TODO: change signature so `value` accepts only the correct type?
```
do you want to address this TODO? (either that or remove the comment)

xnyo replied:
I tried, but I forgot why I used `any` in the first place:

```go
package github.com/grafana/grafana/pkg/cmd/grafana
	imports github.com/grafana/grafana/pkg/cmd/grafana-cli/commands
	imports github.com/grafana/grafana/pkg/cmd/grafana-cli/commands/secretsmigrations
	imports github.com/grafana/grafana/pkg/server
	imports github.com/grafana/grafana/pkg/registry/backgroundsvcs
	imports github.com/grafana/grafana/pkg/services/pluginsintegration/angulardetectorsprovider
	imports github.com/grafana/grafana/pkg/services/pluginsintegration/angularpatternsstore
	imports github.com/grafana/grafana/pkg/services/pluginsintegration/angulardetectorsprovider: import cycle not allowed
```

So I will leave it for a follow-up PR eventually

andresmgot commented on pkg/services/pluginsintegration/cachekvstore/namespacedstore.go:
```diff
@@ -0,0 +1,141 @@
+	storeKeyGetter StoreKeyGetter
+}
+
+// DefaultStoreKeyGetter is the default StoreKeyGetterFunc, which returns the key as-is.
+var DefaultStoreKeyGetter = StoreKeyGetterFunc(func(k string) string {
+	return k
+})
+
+// PrefixStoreKeyGetter returns a StoreKeyGetterFunc that returns the key with a prefix.
+// It is used to prefix all keys in the underlying store with a fixed prefix.
+func PrefixStoreKeyGetter(prefix string) StoreKeyGetterFunc {
+	return func(k string) string {
+		return prefix + k
+	}
+}
+
+// NamespacedStoreOpt is an option for NewNamespacedStore that modifies the store passed to it.
+type NamespacedStoreOpt func(store *NamespacedStore)
+
+// WithStoreKeyGetter sets the StoreKeyGetter to use.
+func WithStoreKeyGetter(g StoreKeyGetter) NamespacedStoreOpt {
+	return func(store *NamespacedStore) {
+		store.storeKeyGetter = g
+	}
+}
```
are all of this necessary to be exported? To simplify it you can just replace `NamespacedStoreOpt` with a `prefix string` and just expose:

```
// NewNamespacedStore creates a new NamespacedStore using the provided underlying KVStore, namespace and optionally a prefix.
func NewNamespacedStore(kv kvstore.KVStore, namespace string, prefix string) plugins.KeyStore {
```

Honestly I think that the prefix is not needed but we can keep it to avoid changes in the database.

andresmgot reviewed (commented):
:+1:  I think this can be simplified even further, let me know if I missed something

xnyo committed:
PR review: removed unused interfaces

xnyo committed:
PR review: Moved marshal into namespacedstore.go

xnyo committed:
PR review: removed storekeygetter

xnyo committed:
Removed unused file cachekvstore.go

xnyo committed:
Renamed NamespacedStore to CacheKvStore

xnyo committed:
removed todo

xnyo commented:
Thanks for the reviews, I have simplified it even further:
- Removed options and allow `prefix` as argument
- Removed `KeyStoreGetter`
I have also renamed `NamespacedStore` to `CacheKvStore` as it's the only implementation now

andresmgot commented on pkg/services/pluginsintegration/angularpatternsstore/store_test.go:
```diff
@@ -59,7 +60,7 @@ func TestAngularPatternsStore(t *testing.T) {
 		})
 
 		t.Run("invalid timestamp stored", func(t *testing.T) {
```
nit: I guess you can remove these test since it's written now in the common implementation

andresmgot reviewed (approved):
LGTM! Thanks!
